File: src/pong.rs
use crate::selector::SelectedGame;
use macroquad::{
    audio::{play_sound, PlaySoundParams, Sound},
    prelude::*,
};

const PADDLE_WIDTH: f32 = 200.0;
const PADDLE_HEIGHT: f32 = 10.0;
const PADDLE_PADDING: f32 = 20.0;
const PADDLE_SPEED: f32 = 180.0;

const BALL_RADIUS: f32 = 5.0;
const DEFAULT_VELOCITY: Vec2 = vec2(170.0, 170.0);

struct Ball {
    velocity: Vec2,
    hitbox: Rect,
}

pub struct Pong {
    window_size: Vec2,
    ball: Ball,
    score: (u32, u32),
    paddle_top: Rect,
    paddle_bottom: Rect,
    countdown_passed: Option<f32>,
    pub sound: Option<Sound>,
}

impl Pong {
    fn reset(&mut self, bottom_scored: bool) {
        let paddle_x = self.window_size.x / 2.0 - PADDLE_WIDTH / 2.0;

        self.paddle_bottom.x = paddle_x;
        self.paddle_top.x = paddle_x;
        self.ball.hitbox = Rect::new(
            self.window_size.x / 2.0,
            if bottom_scored {
                self.paddle_top.y + PADDLE_PADDING
            } else {
                self.paddle_bottom.y - PADDLE_PADDING
            },
            BALL_RADIUS,
            BALL_RADIUS,
        );
        self.ball.velocity = DEFAULT_VELOCITY;

        self.countdown_passed = Some(0.0);
    }
    fn draw(&self) {
        clear_background(BLACK);

        let line_y = self.window_size.y / 2.0;
        let score_x = self.window_size.x / 2.0 - 10.0;

        draw_line(0.0, line_y, self.window_size.x, line_y, 2.0, DARKGRAY);
        draw_text(
            &self.score.1.to_string(),
            score_x,
            line_y - 30.0,
            50.0,
            GRAY,
        );
        draw_text(
            &self.score.0.to_string(),
            score_x,
            line_y + 50.0,
            50.0,
            GRAY,
        );

        draw_circle(self.ball.hitbox.x, self.ball.hitbox.y, BALL_RADIUS, WHITE);

        // draw paddles
        draw_rectangle(
            self.paddle_bottom.x,
            self.paddle_bottom.y,
            PADDLE_WIDTH,
            PADDLE_HEIGHT,
            WHITE,
        );
        draw_rectangle(
            self.paddle_top.x,
            self.paddle_top.y,
            PADDLE_WIDTH,
            PADDLE_HEIGHT,
            WHITE,
        );
    }
}

impl crate::game::Game for Pong {
    fn new() -> Self {
        let win_size = Vec2::new(screen_width(), screen_height());
        let win_size_center = win_size / 2.0;
        let paddle_x = win_size_center.x - PADDLE_WIDTH / 2.0;

        Self {
            ball: Ball {
                velocity: DEFAULT_VELOCITY,
                hitbox: Rect::new(
                    win_size_center.x,
                    win_size_center.y,
                    BALL_RADIUS,
                    BALL_RADIUS,
                ),
            },
            score: (0, 0),
            paddle_top: Rect::new(paddle_x, PADDLE_PADDING, PADDLE_WIDTH, PADDLE_HEIGHT),
            paddle_bottom: Rect::new(
                paddle_x,
                win_size.y - PADDLE_HEIGHT - PADDLE_PADDING,
                PADDLE_WIDTH,
                PADDLE_HEIGHT,
            ),
            window_size: win_size,
            countdown_passed: Some(0.0),
            sound: None,
        }
    }
    fn update(&mut self) -> SelectedGame {
        self.window_size = Vec2::new(screen_width(), screen_height());
        self.paddle_bottom.y = self.window_size.y - PADDLE_HEIGHT - PADDLE_PADDING;
        self.paddle_top.y = PADDLE_PADDING;

        let dt = get_frame_time();

        if is_key_pressed(KeyCode::Escape) {
            return SelectedGame::None;
        }

        if let Some(cp) = self.countdown_passed.as_mut() {
            if *cp >= 3.0 {
                self.countdown_passed = None;
            } else {
                *cp += get_frame_time();
            }
        } else {
            let fixed_paddle_speed = PADDLE_SPEED * dt;
            if is_key_down(KeyCode::A) {
                self.paddle_bottom.x = (self.paddle_bottom.x - fixed_paddle_speed)
                    .clamp(0.0, self.window_size.x - PADDLE_WIDTH);
            }
            if is_key_down(KeyCode::D) {
                self.paddle_bottom.x = (self.paddle_bottom.x + fixed_paddle_speed)
                    .clamp(0.0, self.window_size.x - PADDLE_WIDTH);
            }
            if is_key_down(KeyCode::Left) {
                self.paddle_top.x = (self.paddle_top.x - fixed_paddle_speed)
                    .clamp(0.0, self.window_size.x - PADDLE_WIDTH);
            }
            if is_key_down(KeyCode::Right) {
                self.paddle_top.x = (self.paddle_top.x + fixed_paddle_speed)
                    .clamp(0.0, self.window_size.x - PADDLE_WIDTH);
            }
        }

        match self.countdown_passed {
            None => {
                const AUDIO_PARAMS: PlaySoundParams = PlaySoundParams {
                    looped: false,
                    volume: 0.3,
                };

                self.ball.hitbox.x += self.ball.velocity.x * dt;
                self.ball.hitbox.y += self.ball.velocity.y * dt;

                // screen side collide || side paddle collide
                if self.ball.hitbox.x - BALL_RADIUS <= 0.0
                    || self.ball.hitbox.x + BALL_RADIUS >= self.window_size.x
                    || (self.ball.hitbox.overlaps(&self.paddle_top) && self.ball.hitbox.y <= self.paddle_top.bottom())
                {
                    self.ball.velocity.x *= -1.0;
                    play_sound(self.sound.as_ref().unwrap(), AUDIO_PARAMS);
                }

                // bottom scored
                if self.ball.hitbox.y + BALL_RADIUS <= 0.0 {
                    self.reset(true);
                    self.score.0 += 1;
                }
                // top scored
                if self.ball.hitbox.y + BALL_RADIUS >= self.window_size.y {
                    self.reset(false);
                    self.score.1 += 1;
                }

                // bottom paddle
                let bottom_paddle_collided = self.paddle_bottom.overlaps(&self.ball.hitbox);
                let top_paddle_collided = self.paddle_top.overlaps(&self.ball.hitbox);

                // top paddle
                if bottom_paddle_collided || top_paddle_collided {
                    self.ball.velocity.y *= -1.0;
                    play_sound(self.sound.as_ref().unwrap(), AUDIO_PARAMS);
                }

                self.draw();
            }
            Some(cp) => {
                self.draw();

                let x = self.window_size.x / 2.0 - 30.0;
                let y = self.window_size.y / 2.0 + 28.0;
                let size = 125.0;

                if cp <= 1.0 {
                    draw_text("3", x, y, size, GREEN);
                } else if cp <= 2.0 {
                    draw_text("2", x, y, size, YELLOW);
                } else {
                    draw_text("1", x, y, size, RED);
                }
            }
        }

        SelectedGame::Pong
    }
}

File: src/pong/pong.cpp
#ifndef PONG_C_
#define PONG_C_

#include "raylib.h"
#include "raymath.h"
#include "../const.hpp"
#include "../gamestate.hpp"
#include "../assets.h"

#include "paddle.hpp"
#include "ball.hpp"

INCBIN(pong_hit_sound, "assets/beep.wav");

struct Score {
    uint32_t left;
    uint32_t right;
};

struct Pong {
private:
    float paddle_y;
    float abs_vx;

public:
    Paddle paddle_left;
    Paddle paddle_right;
    Ball ball;
    GameState state;
    float countdown_passed;
    Sound hit_sound;
    Sound score_sound;
    Score score;
    Vector2 window_size;

    float dt;
    float fixed_paddle_speed;
    float line_x, score_y;

    Pong() {
        window_size = vec2(GetScreenWidth(), GetScreenHeight());
        Vector2 center;
        center.x = window_size.x / 2;
        center.y = window_size.y / 2;

        float paddle_y = center.y - PADDLE_HEIGHT / 2;

        ball.radius = BALL_RADIUS;
        ball.pos = center;
        ball.velocity = BALL_DEFAULT_VELOCITY;
        score = (Score) {0, 0};

        paddle_left = rect(
            PADDLE_PADDING, paddle_y, PADDLE_WIDTH, PADDLE_HEIGHT
        );
        paddle_right = rect(
            window_size.x - PADDLE_PADDING - PADDLE_WIDTH, paddle_y,
            PADDLE_WIDTH,
            PADDLE_HEIGHT
        );

        state = GameState_Help;
        countdown_passed = 0.0f;
        hit_sound = LoadSoundFromWave(
            LoadWaveFromMemory(
                ".wav", pong_hit_sound_data, pong_hit_sound_size
            )
        );
        score_sound = LoadSoundFromWave(
            LoadWaveFromMemory(
                ".wav", pong_hit_sound_data, pong_hit_sound_size
            )
        );
        SetSoundVolume(hit_sound, 0.3);
        SetSoundVolume(score_sound, 0.5);
        SetSoundPitch(score_sound, 1.6);
    }

    ~Pong() {
        UnloadSound(hit_sound);
        UnloadSound(score_sound);
    }

    void reset(bool left_scored) {
        paddle_y = (window_size.y / 2) - (PADDLE_HEIGHT / 2);

        paddle_right.y = paddle_y;
        paddle_left.y = paddle_y;
        ball.pos = vec2(
            left_scored?
                PADDLE_PADDING*2 + PADDLE_WIDTH :
                window_size.x - PADDLE_PADDING*2 - PADDLE_WIDTH,
            window_size.y / 2
        );

        abs_vx = fabsf(ball.velocity.x);

        ball.velocity.x = left_scored? abs_vx : -abs_vx;
        ball.velocity.y = GetRandomValue(0, 1)? 1.0f : -1.0f;

        countdown_passed = 0.0f;
    }

    void draw_help() {
        BeginDrawing();
        ClearBackground(BLACK);
        // TODO
    }

    void draw() {
        BeginDrawing();
        ClearBackground(DARKGRAY);

        DrawTextD("Pong is work in progress :P", window_size.x / 2 - 300, window_size.y / 2 - 50, 50.0, GOLD);
        return;

        line_x = window_size.x / 2.0f;
        score_y = window_size.y / 2.0f - 25.0f;

        BeginDrawing();
        ClearBackground(BLACK);

        DrawLine(line_x, window_size.y, line_x, 0.0f, DARKGRAY);
        
        g_sprintf("%u", score.left);
        DrawTextD(
            global_text_buf, line_x - 75.0,
            score_y,
            50.0, GRAY
        );
        g_sprintf("%u", score.right);
        DrawTextD(
            global_text_buf, line_x + 50.0,
            score_y,
            50.0, GRAY
        );

        DrawCircleV(ball.pos, ball.radius, WHITE);

        DrawRectangleRec(
            paddle_left,
            RAYWHITE
        );
        DrawRectangleRec(
            paddle_right,
            RAYWHITE
        );
    }

    SelectedGame update() {
        window_size = vec2(GetScreenWidth(), GetScreenHeight());
        paddle_right.x = window_size.x - PADDLE_WIDTH - PADDLE_PADDING;

        this->draw();

        dt = GetFrameTime();

        if (IsKeyPressed(KEY_ESCAPE)) {
            return Selected_None;
        }
        return Selected_PONG; // NOTE remove later :p

        if (IsKeyPressed(KEY_H)) {
            state = (state == GameState_Help)? GameState_Running : GameState_Help;
            this->draw();
            return Selected_PONG;
        }

        if (state == GameState_Help) {
            this->draw_help();
            return Selected_PONG;
        }

        if (IsKeyPressed(KEY_R)) {
            score.left = 0;
            score.right = 0;
            state = GameState_Running;

            this->reset(GetRandomValue(0, 1));
        }
        if (IsKeyPressed(KEY_SPACE)) {
            state = (state == GameState_Running) ? GameState_Paused : GameState_Running;
            this->draw();
            return Selected_PONG;
        }

        if (state == GameState_Paused) {
            this->draw();
            DrawTextD(
                "PAUSED",
                3, window_size.y - 25,
                25.0, GREEN
            );
            return Selected_PONG;
        }

        if (countdown_passed >= 0.0f) {
            countdown_passed =
                (countdown_passed >= 3.0f)?
                -1.0f : countdown_passed + dt;
        } else {
            fixed_paddle_speed = PADDLE_SPEED * dt;
            if (IsKeyDown(KEY_W)) {
                paddle_left.y = Clamp(
                    paddle_left.y - fixed_paddle_speed,
                    0.0f,
                    window_size.y - PADDLE_HEIGHT
                );
            }
            if (IsKeyDown(KEY_S)) {
                paddle_left.y = Clamp(
                    paddle_left.y + fixed_paddle_speed,
                    0.0f,
                    window_size.y - PADDLE_HEIGHT
                );
            }
            if (IsKeyDown(KEY_UP)) {
                paddle_right.y = Clamp(
                    paddle_right.y - fixed_paddle_speed,
                    0.0f,
                    window_size.y - PADDLE_HEIGHT
                );
            }
            if (IsKeyDown(KEY_DOWN)) {
                paddle_right.y = Clamp(
                    paddle_right.y + fixed_paddle_speed,
                    0.0f,
                    window_size.y - PADDLE_HEIGHT
                );
            }

        }

        if (countdown_passed < 0.0f) {
            ball.pos.x += ball.velocity.x * dt;
            ball.pos.y += ball.velocity.y * dt;

            this->handle_collision();
            this->draw();
        } else {
            this->draw();

            static float x, y, size = 125.0;
            x = window_size.x / 2.0f - 30.0f;
            y = window_size.y / 2.0f- 70.0f;

            if (countdown_passed <= 1.0f)
                DrawTextD(
                    "3", x, y, size, GREEN
                );
            else if (countdown_passed <= 2.0f)
                DrawTextD(
                    "2", x, y, size, YELLOW
                );
            else
                DrawTextD(
                    "1", x + 12, y, size, RED
                );
        }
        return Selected_PONG;
    }

    void handle_collision() {
        // TODO
    }
};

#endif